test/common/upstream/cluster_manager_impl_test.cc

```diff
@@ -2442,14 +2449,226 @@ TEST_F(ClusterManagerImplTest, UseTcpWithCustomDnsResolver) {
   )EOF";
 
   std::shared_ptr<Network::MockDnsResolver> dns_resolver(new Network::MockDnsResolver());
-  // `true` here stands for using tcp
-  EXPECT_CALL(factory_.dispatcher_, createDnsResolver(_, true)).WillOnce(Return(dns_resolver));
+  envoy::config::core::v3::DnsResolverOptions dns_resolver_options;
+  EXPECT_CALL(factory_.dispatcher_, createDnsResolver(_, _))
+      .WillOnce(DoAll(SaveArg<1>(&dns_resolver_options), Return(dns_resolver)));
+
+  Network::DnsResolver::ResolveCb dns_callback;
+  Network::MockActiveDnsQuery active_dns_query;
+  EXPECT_CALL(*dns_resolver, resolve(_, _, _))
+      .WillRepeatedly(DoAll(SaveArg<2>(&dns_callback), Return(&active_dns_query)));
+  create(parseBootstrapFromV3Yaml(yaml));
+  // `true` here means use_tcp_for_dns_lookups is being set via bootstrap config
+  EXPECT_EQ(true, dns_resolver_options.has_use_tcp_for_dns_lookups());
+  // `true` here means use_tcp_for_dns_lookups is set to true
+  EXPECT_EQ(true, dns_resolver_options.use_tcp_for_dns_lookups().value());
+  factory_.tls_.shutdownThread();
+}
+
+// Test that custom DNS resolver with UDP lookups is used, when custom resolver is configured
+// per cluster and `use_tcp_for_dns_lookups` is specified as true but is overridden
+// by dns_resolver_options.use_tcp_for_dns_lookups which is set as false.
+TEST_F(ClusterManagerImplTest, UseUdpWithCustomDnsResolverDeprecatedFieldOverridden) {
+  const std::string yaml = R"EOF(
+  static_resources:
+    clusters:
+    - name: cluster_1
+      use_tcp_for_dns_lookups: true
+      dns_resolver_options:
+        use_tcp_for_dns_lookups: false
```

Comment:
We should fail the config if this is ever observed.  Enforce a correct config instead of trying to "guess" what the user intended.

Reply:
I have never thought about failing the config if both these conflicting values are set. In the code logic, preference is given for _use_tcp_for_dns_lookups_ via _dns_resolver_options_ rather than bare _use_tcp_for_dns_lookups_.  So any value to _**dns_resolver_options.use_tcp_for_dns_lookups**_ will override _**use_tcp_for_dns_lookups**_. 

In other words _**use_tcp_for_dns_lookups**_ will come in to effect only if _**dns_resolver_options.use_tcp_for_dns_lookups**_ is not set any bool value.


Let me know if you strongly suggest that we should fail the config if both these fields are set with conflicting values. I will have to look into how to NACK such config.